azure/scope/strategies/machinepool_deployments/machinepool_deployment_strategy.go

```diff
@@ -233,12 +234,17 @@ func getFailedMachines(machinesByProviderID map[string]infrav1exp.AzureMachinePo
 	return machines
 }
 
+// getDeletingMachines is responsible for identifying machines whose VMs are in an active state of deletion
+// but whose corresponding AzureMachinePoolMachine resource has not yet been marked for deletion.
 func getDeletingMachines(machinesByProviderID map[string]infrav1exp.AzureMachinePoolMachine) []infrav1exp.AzureMachinePoolMachine {
 	var machines []infrav1exp.AzureMachinePoolMachine
 	for _, v := range machinesByProviderID {
-		// ready status, with provisioning state Succeeded, and not marked for delete
+		// provisioning state is Deleting
 		if v.Status.ProvisioningState != nil && *v.Status.ProvisioningState == infrav1.Deleting {
-			machines = append(machines, v)
+			// If we've already marked the DeletionTimestamp then we have no work to do
+			if v.DeletionTimestamp.IsZero() {
```

Comment:
```suggestion
			if v.DeletionTimestamp.IsZero() || (v.Status.ProvisioningState != nil && *v.Status.ProvisioningState == infrav1.Deleting) {
```

I was thinking about this last night and in the case where the AMPM is not marked for deletion, but the AMPM last reconcile observed the VM is being deleted in Azure, this would ensure AMP reconcile loop doesn't delete an extra machine. 

This is an edge case optimization, but I think it might be worth the change.

Reply:
We shouldn't get to this point in the flow if we have any "Azure is deleting a VM but capz doesn't know about it yet", because we track those VMs when we call `getDeletingMachines()` earlier; if `getDeletingMachines()` returns a non-empty array we will short-circuit before we introspect the result from `readyMachines()`.

However...

The order of evaluation could change in the future and so to make this more reliable going forward we should probably take this into account when preparing our set of Ready machines. I'll make this change.

Reply:
I guess what I was saying is that I'd prefer the if statement combined like what you did with ready machines.

Reply:
Sorry, I was totally confused here.